aws/rust-runtime/aws-config/src/wasi_adapter/mod.rs

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! HTTP Service Adapter used in WASI environment
+
+use aws_smithy_http::body::SdkBody;
+use aws_smithy_http::result::ConnectorError;
+use bytes::Bytes;
+use http::{Request, Response};
+use std::task::{Context, Poll};
+use tower::Service;
+use wasi_preview2_prototype::http_client::DefaultClient;
+
+#[derive(Default, Debug, Clone)]
+pub(crate) struct Adapter {}
+
+impl Service<Request<SdkBody>> for Adapter {
+    type Response = Response<SdkBody>;
+
+    type Error = ConnectorError;
+
+    #[allow(clippy::type_complexity)]
+    type Future = std::pin::Pin<
+        Box<dyn std::future::Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>,
+    >;
+
+    fn poll_ready(&mut self, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn call(&mut self, req: Request<SdkBody>) -> Self::Future {
+        println!("Adapter: sending request...");
+        let client = DefaultClient::new(None);
+        // Right now only synchronous calls can be made through WASI
+        let fut = client.handle(req.map(|body| match body.bytes() {
+            Some(value) => Bytes::copy_from_slice(value),
+            None => Bytes::new(),
+        }));
```

Comment:
seems like the chances of deadlock here are pretty high...

Reply:
i believe this will eventually be solved when threading is supported fully within WebAssembly/WASI. for the http proposal, it will most likely happen only with WASI preview 3 (more details [here](https://github.com/WebAssembly/WASI/issues/515)).